From 6504a8852440fecde7529646276fe8d7dffd859c Mon Sep 17 00:00:00 2001
From: Nigel Jones <nigel.l.jones+git@gmail.com>
Date: Fri, 24 May 2019 12:09:09 +0100
Subject: [PATCH 3/4] Fixup Admin services responses void and success

Signed-off-by: Nigel Jones <nigel.l.jones+git@gmail.com>
---
 .../admin/server/spring/OpenMetadataAdminResource.java   | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/open-metadata/src/main/java/org/apache/atlas/openmetadata/admin/server/spring/OpenMetadataAdminResource.java b/open-metadata/src/main/java/org/apache/atlas/openmetadata/admin/server/spring/OpenMetadataAdminResource.java
index 6fa324510..1469ffc25 100644
--- a/open-metadata/src/main/java/org/apache/atlas/openmetadata/admin/server/spring/OpenMetadataAdminResource.java
+++ b/open-metadata/src/main/java/org/apache/atlas/openmetadata/admin/server/spring/OpenMetadataAdminResource.java
@@ -24,7 +24,8 @@ import org.odpi.openmetadata.adminservices.OMAGServerAdminServices;
 import org.odpi.openmetadata.adminservices.OMAGServerOperationalServices;
 import org.odpi.openmetadata.adminservices.configuration.properties.*;
 import org.odpi.openmetadata.adminservices.rest.OMAGServerConfigResponse;
-import org.odpi.openmetadata.adminservices.rest.VoidResponse;
+import org.odpi.openmetadata.adminservices.rest.SuccessMessageResponse;
+import org.odpi.openmetadata.commonservices.ffdc.rest.VoidResponse;
 import org.odpi.openmetadata.frameworks.connectors.properties.beans.*;
 
 import org.slf4j.Logger;
@@ -446,8 +447,8 @@ public class OpenMetadataAdminResource
      * OMAGConfigurationErrorException there is a problem using the supplied configuration.
      */
     @RequestMapping(method = RequestMethod.POST, path = "/instance")
-    public VoidResponse activateWithStoredConfig(@PathVariable String userId,
-                                                 @PathVariable String serverName)
+    public SuccessMessageResponse activateWithStoredConfig(@PathVariable String userId,
+                                                           @PathVariable String serverName)
     {
         return operationsAPI.activateWithStoredConfig(userId, serverName);
     }
@@ -466,7 +467,7 @@ public class OpenMetadataAdminResource
      * OMAGConfigurationErrorException there is a problem using the supplied configuration.
      */
     @RequestMapping(method = RequestMethod.POST, path = "/instance/configuration")
-    public VoidResponse activateWithSuppliedConfig(@PathVariable String           userId,
+    public SuccessMessageResponse activateWithSuppliedConfig(@PathVariable String           userId,
                                                    @PathVariable String           serverName,
                                                    @RequestParam OMAGServerConfig configuration)
     {
-- 
2.21.0

