From aa57e31c3c0e0fd7bc1f432354eba40119292318 Mon Sep 17 00:00:00 2001
From: Nigel Jones <nigel.l.jones+git@gmail.com>
Date: Mon, 24 Dec 2018 14:48:46 +0000
Subject: [PATCH] #492 additional patches for egeria newer spring fw

Signed-off-by: Nigel Jones <nigel.l.jones+git@gmail.com>
---
 .../atlas/web/filters/NullServletContext.java | 31 +++++++++++++++++++
 1 file changed, 31 insertions(+)

diff --git a/webapp/src/main/java/org/apache/atlas/web/filters/NullServletContext.java b/webapp/src/main/java/org/apache/atlas/web/filters/NullServletContext.java
index 25621c6bb..3c506465b 100644
--- a/webapp/src/main/java/org/apache/atlas/web/filters/NullServletContext.java
+++ b/webapp/src/main/java/org/apache/atlas/web/filters/NullServletContext.java
@@ -297,5 +297,36 @@ public class NullServletContext implements ServletContext {
         return null;
     }
 
+    public void setResponseCharacterEncoding(String encoding)
+    {
+    }
+
+    public String getResponseCharacterEncoding()
+    {
+	return null;
+    }
+
+    public void setRequestCharacterEncoding(String encoding)
+    {
+    }
+
+    public String getRequestCharacterEncoding()
+    {
+	return null;
+    }
+
+    public void setSessionTimeout(int timout)
+    {
+    }
+
+    public int getSessionTimeout()
+    {
+        return 0;
+    }
+
+    public javax.servlet.ServletRegistration.Dynamic addJspFile(String jspName, String jspFile)
+    {
+	return null;
+    }
 
 }
-- 
2.20.1

